src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ConnectionStringTopologyProvider.java

```diff
@@ -0,0 +1,283 @@
+ * other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package com.mysql.cj.jdbc.ha.plugins;
+
+import com.mysql.cj.Messages;
+import com.mysql.cj.conf.ConnectionUrl;
+import com.mysql.cj.conf.ConnectionUrlParser;
+import com.mysql.cj.conf.HostInfo;
+import com.mysql.cj.conf.PropertyKey;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.log.Log;
+import com.mysql.cj.util.IpAddressUtils;
+import com.mysql.cj.util.StringUtils;
+
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.IP_ADDRESS;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.OTHER;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.RDS_CUSTOM_CLUSTER;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.RDS_INSTANCE;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.RDS_PROXY;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.RDS_READER_CLUSTER;
+import static com.mysql.cj.jdbc.ha.plugins.RdsUrlType.RDS_WRITER_CLUSTER;
+
+public class ConnectionStringTopologyProvider {
+  private final Pattern auroraDnsPattern =
+      Pattern.compile(
+          "(.+)\\.(proxy-|cluster-|cluster-ro-|cluster-custom-)?([a-zA-Z0-9]+\\.[a-zA-Z0-9\\-]+\\.rds\\.amazonaws\\.com)",
+          Pattern.CASE_INSENSITIVE);
+  private final Pattern rdsInstanceDnsPattern =
+      Pattern.compile(
+          "(.+)\\.(proxy-|cluster-|cluster-ro-|cluster-custom-){0}([a-zA-Z0-9]+\\.[a-zA-Z0-9\\-]+\\.rds\\.amazonaws\\.com)",
+          Pattern.CASE_INSENSITIVE);
+  private final Pattern auroraCustomClusterPattern =
+      Pattern.compile(
+          "(.+)\\.(cluster-custom-[a-zA-Z0-9]+\\.[a-zA-Z0-9\\-]+\\.rds\\.amazonaws\\.com)",
+          Pattern.CASE_INSENSITIVE);
+  private final Pattern auroraProxyDnsPattern =
+      Pattern.compile(
+          "(.+)\\.(proxy-[a-zA-Z0-9]+\\.[a-zA-Z0-9\\-]+\\.rds\\.amazonaws\\.com)",
+          Pattern.CASE_INSENSITIVE);
+  private final Log logger;
+  private final int failoverConnectTimeoutMs;
```

Comment:
Do we use them?

Reply:
The logger we use in this class to throw an error if the host pattern setting is invalid. I can convert the failoverConnectTimeout and socketTimeout to function parameters, they are stored in the cluster instance template. However, I'm not sure if its necessary to store these values in the cluster instance template or if they are used anywhere, which is maybe what your question was getting at